# Fitts' Law – UX Performance Modell

Autor: Luca Honegger, lh@sinnhaft.ch

Quelle: Dies ist eine Python-Portierung der UXToolbox in R von Mohsen Rafiei, https://github.com/mohsen-rafiei/UXtoolbox


## 🧠 Was macht diese Funktion?

Die Funktion `fitts_law()` sagt vorher, **wie lange ein Mensch braucht**, um mit einem Interface-Element zu interagieren – zum Beispiel einen Button zu klicken.

Sie basiert auf dem bekannten **Fitts'schen Gesetz**, das in der Mensch-Computer-Interaktion verwendet wird, um die Effizienz von Benutzerschnittstellen zu analysieren.


## 🔍 Was wird berechnet?

Die Funktion verwendet die Formel:

$$
\text{Bewegungszeit} = a + b \cdot \log_2\left(\frac{\text{Distanz}}{\text{Breite}} + 1\right)
$$

Dabei gilt:
* **Distanz**: Abstand vom Startpunkt zum Ziel (z. B. Mausweg)
* **Breite**: Breite des Ziels (z. B. Button-Größe)
* **a, b**: empirisch bestimmte Regressionsparameter

Optional kann auch eine **gewichtete Variante** verwendet werden, um realistischere UX-Szenarien zu modellieren (z. B. bei kognitiven/motorischen Verzögerungen).


## 🧪 Beispiel aus dem UX-Alltag

Stell dir vor, du vergleichst zwei UI-Designs:

* In Design A liegt ein Button 100 px entfernt und ist 10 px breit
* In Design B liegt ein Button 300 px entfernt und ist 30 px breit

Mit `fitts_law([100, 300], [10, 30])` bekommst du:

* Die **geschätzte Bewegungszeit**, die Nutzer:innen für jede Variante benötigen
* Damit kannst du Designs **objektiv vergleichen**, z. B. in Usability-Tests oder Prototypen


## 🎯 Warum ist das wichtig?

Fitts' Law ist ein zentraler Baustein der **UX-Metriken** – und wird in vielen Standards verwendet (z. B. ISO 9241). Es hilft bei:

* Bewertung der **Effizienz von User Interfaces**
* Planung von **interaktiven Elementen** (Größe, Abstand)
* Erkennung von **Barrieren bei motorischer Interaktion**

Mit dieser Funktion kannst du im Unterricht oder Projekt direkt nachvollziehen, wie **Interaktionsdesign messbar und optimierbar** wird.


## `fitts_law`

In [6]:
import numpy as np

def fitts_law(distance, width, a=0.2, b=0.1, weighted=False):
    """
    Fitts'sches Gesetz zur Vorhersage der UX-Leistungsfähigkeit.

    Berechnet die Bewegungszeit basierend auf der Distanz zum Ziel und der Zielbreite.
    Unterstützt sowohl das Standardmodell als auch eine gewichtete Variante.

    Parameter:
    - distance (array-like): Distanz(en) zum Ziel (in Pixel oder cm).
    - width (array-like): Breite(n) des Ziels (in Pixel oder cm).
    - a (float): Regressionskonstante (Standardwert: 0.2).
    - b (float): Regressionskoeffizient (Standardwert: 0.1).
    - weighted (bool): Ob eine gewichtete Schwierigkeit verwendet werden soll (Standard: False).

    Rückgabe:
    - numpy.ndarray: Vorhergesagte Bewegungszeiten.
    """
    # Umwandlung der Eingaben in NumPy-Arrays (für Vektoroperationen)
    distance = np.asarray(distance)
    width = np.asarray(width)

    # Berechnung des Schwierigkeitsgrads (Index of Difficulty)
    if weighted:
        # Gewichtete Variante (z. B. zur realitätsnäheren Modellierung)
        index_of_difficulty = np.log2((distance / width) + 1) * 1.1
    else:
        # Standardformel gemäß Fitts' Law
        index_of_difficulty = np.log2((distance / width) + 1)

    # Berechnung der Bewegungszeit nach Fitts' Gesetz
    movement_time = a + (b * index_of_difficulty)
    return movement_time


## Anwendungsbeispiel

In [34]:
# Beispielwerte für Distanz und Breite des Ziels in Pixel
distanzen = [100, 200, 500]
breiten = [10, 32, 64]

# Bewegungsgeschwindigkeit mit Standardmodell
print("Standardmodell:", fitts_law(distanzen, breiten))

# Bewegungsgeschwindigkeit mit gewichteter Variante
print("Gewichtetes Modell:", fitts_law(distanzen, breiten, weighted=True))


Standardmodell: [0.54594316 0.4857981  0.51395514]
Gewichtetes Modell: [0.58053748 0.51437791 0.54535065]


### 📊 Erklärung und Interpretation

#### 🔹 Standardmodell: `[0.5459, 0.4858, 0.5140]`

* Diese Werte sind die **geschätzten Bewegungszeiten** für drei verschiedene Interaktionsszenarien mit unterschiedlichen Kombinationen aus Ziel-Distanz und Ziel-Breite.

* Die Berechnung erfolgt auf Basis der klassischen Fitts’schen Formel:

  $$
  MT = a + b \cdot \log_2\left(\frac{D}{W} + 1\right)
  $$

* Interpretation der einzelnen Werte:

  * **0.5459 s** → mittelschwere Interaktion mit relativ hoher Schwierigkeit (z. B. kleiner Button bei größerem Abstand)
  * **0.4858 s** → etwas leichtere Interaktion (größeres Ziel oder geringere Entfernung)
  * **0.5140 s** → mittlerer Schwierigkeitsgrad

* Diese Werte sind als **Baseline** für Bewegungszeit-Schätzungen ohne Zusatzfaktoren wie Stress oder Ablenkung zu verstehen.


#### 🔸 Gewichtetes Modell: `[0.5805, 0.5144, 0.5454]`

* Hier wird der **Index of Difficulty mit 1.1 multipliziert**, um eine realistischere UX-Situation abzubilden.

* Die resultierenden Werte sind durchweg etwas höher:

  * **0.5805 s** → \~6.3 % mehr als das Standardmodell
  * **0.5144 s** → ebenfalls leicht erhöht
  * **0.5454 s** → realistische Anpassung für mittlere Schwierigkeit

* Die Gewichtung berücksichtigt z. B.:

  * Motorische Ungenauigkeiten (z. B. zitternde Hände)
  * Kognitive Belastung (z. B. Ablenkung)
  * Interface-Kontext (z. B. auf Touchscreens oder bei Lichtreflexion)



### 📌 Fazit

| Modell                 | Bewegungszeit (s)     | Interpretation                                            |
| ---------------------- | --------------------- | --------------------------------------------------------- |
| **Standardmodell**     | 0.546 / 0.486 / 0.514 | Mathematische Basiszeit je nach Interaktionsschwierigkeit |
| **Gewichtetes Modell** | 0.581 / 0.514 / 0.545 | Realistischere Schätzung unter erschwerten Bedingungen    |

* Der Unterschied liegt je nach Fall bei **\~5–7 %** – genug, um UX-Designentscheidungen zu beeinflussen.
* Gerade bei **engen Zeitbudgets**, **Assistenzsystemen** oder **mobilen Geräten** kann das entscheidend sein.
* Diese Werte unterstützen dich dabei, **UX quantitativ zu bewerten** und Interfaces gezielt zu verbessern.
